

```{r}
# Set the working directory
setwd("/Users/tejsharm/Desktop/MECOM_Project/Mecom_paper/MECOM_paper_files/")

```


```{r}
# Update BiocManager and check Bioconductor version
BiocManager::install(version = "3.18")
BiocManager::valid()

# Function to install and load CRAN and Bioconductor packages
install_and_load <- function(cran_packages = NULL, bioc_packages = NULL) {
  if (!is.null(cran_packages)) {
    for (pkg in cran_packages) {
      if (!requireNamespace(pkg, quietly = TRUE)) {
        install.packages(pkg, dependencies = TRUE)
      }
      library(pkg, character.only = TRUE, quietly = TRUE)
    }
  }
  
  if (!is.null(bioc_packages)) {
    if (!requireNamespace("BiocManager", quietly = TRUE)) {
      install.packages("BiocManager")
    }
    for (pkg in bioc_packages) {
      if (!requireNamespace(pkg, quietly = TRUE)) {
        BiocManager::install(pkg, ask = FALSE)
      }
      library(pkg, character.only = TRUE, quietly = TRUE)
    }
  }
}

# Define package lists
cran_packages <- c(
  "readr", "readxl", "tidyr", "dplyr", "edgeR", "plyr", "pathview", "tidyverse", 
  "tidygraph", "ggraph", "igraph", "GSEABase", "magrittr", "AnnotationDbi", 
  "Biostrings", "ggplot2", "pheatmap", "org.Hs.eg.db", "reshape", "reshape2", 
  "grid", "gridExtra", "wesanderson", "GOplot", "plotly", "clusterProfiler", 
  "RColorBrewer", "GSVA", "survminer", "ggVennDiagram", "colorspace", 
  "VennDiagram", "msigdbr", "ggrepel", "ggridges", "cowplot", "netgsa", "graphite"
)

install.packages("readxl", dependencies = TRUE)

```


```{r}
library(openxlsx)
library(plyr)
library(pathview)
library(tidyverse)
library(tidygraph)
library(ggraph)
library(igraph)
library(GSEABase)
library(magrittr)
library(AnnotationDbi)
library(Biostrings)
library(pheatmap)
library(org.Hs.eg.db)
library(reshape)
library(reshape2)
library(grid)
library(gridExtra)
library(wesanderson)
library(GOplot)
library(plotly)
library(clusterProfiler)
library(RColorBrewer)
library(GSVA)
library(survminer)
library(ggVennDiagram)
library(colorspace)
library(VennDiagram)
library(msigdbr)
library(ggrepel)
library(ggridges)
library(cowplot)
library(netgsa)
library(graphite)
library(EnsDb.Hsapiens.v75)
library(Homo.sapiens)

# Extract gene annotations
edb <- EnsDb.Hsapiens.v75
gene_annotations <- genes(edb, columns = c("gene_id", "symbol"))
gene_annotations_df <- as.data.frame(gene_annotations)
gene_annotations_df$ensgene <- gene_annotations_df$gene_id

# Convert gene_id to character to match the type in counts
gene_annotations_df$gene_id <- as.character(gene_annotations_df$gene_id)
head(gene_annotations_df)
```


```{r}
library(edgeR)
# Load tsv file (this file is from the pipeline output)
grch37_counts <- read_tsv("/Users/tejsharm/Library/CloudStorage/Box-Box/Adelaiye-Ogala_Remi/RT2867/RT2867_Results/star_rsem/rsem.merged.gene_counts.tsv")
counts <- grch37_counts[,-2]

str(counts$gene_id)

gene_annotations_df <- dplyr::select(gene_annotations_df, gene_id, ensgene, symbol)

counts_annotated <- dplyr::left_join(counts, gene_annotations_df, by = "gene_id")

head(counts_annotated)

# Import samplesheet
sampleIDs <- read.xlsx("Sample_Sheet.xlsx") 

# Build dGE list, Create sample Dataframe
targets <- data.frame(manifestID = colnames(counts[2:30])) %>% 
  left_join(sampleIDs, by = c("manifestID" = "Manifest_ID")) 

# Build DGEList object for differential expression analysis
dge_list_grch37 <- DGEList(counts = as.matrix(counts_annotated[2:30]), 
                           genes = counts_annotated[, c("gene_id", "ensgene", "symbol")],
                           samples = targets)
dge_list_grch37
```


```{r}
library(dplyr)         
library(tidyr)        
library(readr)

keep <- rowSums(edgeR::cpm(dge_list_grch37)>1) >= 3
dge_list_grch37 <- dge_list_grch37[keep, , keep.lib.sizes=FALSE]
rnaData <- dge_list_grch37[filterByExpr(dge_list_grch37, group = "group"),]
rnaData <- calcNormFactors(rnaData)

mean_cpm_df <- edgeR::cpm(rnaData, log = T) %>% 
  as.data.frame() %>% 
  cbind(rnaData$genes[,1]) %>%
  cbind(rnaData$genes[,3]) %>%
  pivot_longer(contains("REP"),
               names_to = "file_name",
               values_to = "cpm")  

sampleData <- rnaData$samples %>% 
  as.data.frame()
mean_cpm_df <- mean_cpm_df %>%
    left_join(sampleData, c("file_name" = "manifestID")) %>% 
  group_by(Sample_ID, `rnaData$genes[, 1]`, `rnaData$genes[, 3]`) %>% 
  summarise(cpm = mean(cpm)) %>% 
  ungroup() %>% 
  pivot_wider(names_from = Sample_ID,
              values_from = cpm)

mean_cpm_df  <-mean_cpm_df  %>% rename(Geneid = `rnaData$genes[, 1]`, Symbol = `rnaData$genes[, 3]`)
mean_cpm_matrix <- as.matrix(mean_cpm_df[,-1])
rownames(mean_cpm_matrix) <- mean_cpm_df$Geneid
mean_cpm_matrix %>% as.data.frame(mean_cpm_matrix) %>% write_csv("mean_countPM.csv")
```


```{r}
library(stats)
library(ggplot2)    
library(ggrepel)

y <- as.data.frame(mean_cpm_matrix[,c(2:11)])
y <- as.data.frame(apply(y, 2, as.numeric))
pca_mean <- prcomp(t(y))
summary(pca_mean)
scores_rnaData = as.data.frame(pca_mean$x)
p <- scores_rnaData[,1:2]
names <- rownames(p)
p2 <- cbind(names, p) 
rownames(p2) <- NULL

ggplot(p2, aes(PC1, PC2, color, label = names))+
  geom_point(aes(shape = 21, fill = names,  color = names, stroke = 2), size = 5)+
  scale_shape_identity()+
  scale_color_manual(values = c("goldenrod3", "darkolivegreen", "red", "blue4", "purple4",
                                "darkgreen", "firebrick4", "cyan","tomato", "plum4"))+
  scale_fill_manual(values = c("C42B_Control" = "goldenrod3", "C42B_ER_Control" = "darkolivegreen",
                               "C42B_ER_shMECOM" = "red", "C42B_shMECOM" = "blue4", 
                               "CWR_R1_Control" = "purple4", "CWR_R1_ER_Control" ="darkgreen",
                               "CWR_R1_ER_shMECOM" ="firebrick4", "CWR_R1_shMECOM"="cyan",
                               "LREX_Control"="tomato","LREX_shMECOM"="plum4"))+
  xlab("PC1")+ ylab("PC2")+
  theme_bw()+
  geom_text_repel(aes(label = names), min.segment.length = 0, 
                  verbose = TRUE, max.time = 5, max.iter = Inf, max.overlaps = Inf, size = 3,
                  nudge_x = 0.15, direction = "both", hjust = 0)+
  xlim(-50, 110) + ylim(-110, 50)+
  ggtitle("PCA plot")+
  theme(legend.position="none")
```


```{r}
library(readr)
condition <- dge_list_grch37$samples$Sample_ID
design = model.matrix(~0 + condition)
colnames(design) <- str_remove_all(colnames(design), "condition")
dge_list <- estimateDisp(dge_list_grch37, design = design)
fit <- glmQLFit(dge_list, design)
cont_mat <- makeContrasts(C42B_shMECOM_v_control = C42B_shMECOM - C42B_Control,
                          C42B_ER_shMECOM_v_control = C42B_ER_shMECOM - C42B_ER_Control,
                          C42B_ER_v_C42B = C42B_ER_Control - C42B_Control, 
                          CWRR1_shMECOM_v_control = CWR_R1_shMECOM - CWR_R1_Control,
                          CWRR1_ER_shMECOM_v_control = CWR_R1_ER_shMECOM - CWR_R1_ER_Control,
                          CWRR1_ER_v_CWRR1 = CWR_R1_ER_Control -CWR_R1_Control,
                          LREX_shMECOM_v_control = LREX_shMECOM - LREX_Control,
                          levels = design)

topTable <- data.frame()
for(contrast in 1:length(colnames(cont_mat))){
  topTable <- topTable %>% 
    bind_rows(topTags(glmQLFTest(fit, contrast = cont_mat[,contrast]), n = Inf)$table %>% 
                mutate(contrast_id = colnames(cont_mat)[contrast]))
}

data <- topTable %>% 
  mutate(
    Expression = case_when(logFC >= 0.5 & PValue <= 0.05 ~ "Up-regulated",
                           logFC <= -0.5 & PValue <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged")
  )

head(data)

write_csv(data, file = "differential_gene_expression.csv")
```


```{r}
top <- 10 
top_genes_C42B_ER_shMECOM <- bind_rows(
  data%>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")) %>% 
    dplyr::filter(Expression == 'Up-regulated') %>% 
    arrange(PValue, desc(abs(logFC)))%>%
    top_n(10, -log10(PValue)),
  data %>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")) %>%
    dplyr::filter(Expression == 'Down-regulated') %>% 
    arrange(PValue, desc(abs(logFC)))%>%
    top_n(10, -log10(PValue))
)

write_csv(top_genes_C42B_ER_shMECOM, "top_genes_C42B_ER_shMECOM_v_control.csv")

top_genes_CWRR1_shMECOM <- bind_rows(
  data%>% dplyr::filter(contrast_id %in% c("CWRR1_shMECOM_v_control")) %>% 
    dplyr::filter(Expression == 'Up-regulated') %>% 
    arrange(PValue, desc(abs(logFC)))%>%
    top_n(10, -log10(PValue)),
  data %>% dplyr::filter(contrast_id %in% c("CWRR1_shMECOM_v_control")) %>%
    dplyr::filter(Expression == 'Down-regulated') %>% 
    arrange(PValue, desc(abs(logFC)))%>%
    top_n(10, -log10(PValue))
)

write_csv(top_genes_CWRR1_shMECOM, "top_genes_CWRR1_shMECOM_v_control.csv")


top_genes_LREX_shMECOM <- bind_rows(
  data%>% dplyr::filter(contrast_id %in% c("LREX_shMECOM_v_control")) %>% 
    dplyr::filter(Expression == 'Up-regulated') %>% 
    arrange(PValue, desc(abs(logFC)))%>%
    top_n(10, -log10(PValue)),
  data %>% dplyr::filter(contrast_id %in% c("LREX_shMECOM_v_control")) %>%
    dplyr::filter(Expression == 'Down-regulated') %>% 
    arrange(PValue, desc(abs(logFC)))%>%
    top_n(10, -log10(PValue))
)

write_csv(top_genes_LREX_shMECOM, "top_genes_LREX_shMECOM_v_control.csv")


csv_files <- c("top_genes_LREX_shMECOM_v_control.csv",
              "top_genes_C42B_ER_shMECOM_v_control.csv",
              "top_genes_CWRR1_shMECOM_v_control.csv")

All_Up_Down_Data <- csv_files %>% 
  lapply(read_csv) %>%
  bind_rows()  


write_csv(All_Up_Down_Data, "All_Up_Down_Data.csv")

# Selected contrast IDs
selected_contrast_ids <- c("CWRR1_shMECOM_v_control", "C42B_ER_shMECOM_v_control", "LREX_shMECOM_v_control")

# Filter all up-regulated and down-regulated genes for the selected contrast IDs
all_regulated_genes <- data %>%
  dplyr::filter(contrast_id %in% selected_contrast_ids, Expression %in% c("Up-regulated", "Down-regulated"))

# Save the filtered data as a CSV file
all_regulated_genes %>% write_csv("All_UpandDown_regulatedGenes_Selected_Contrasts.csv")
```


```{r}
p_value <- 0.05  
q_value <- 0.05 
logFC   <- 1.5

# Ensure genes of interest are present in `data`
genes_C42B_ER <- data %>%
  dplyr::filter(contrast_id == "C42B_ER_shMECOM_v_control", symbol %in% genes_of_interest_C42B_ER)

# Combine top genes and genes of interest
genes_to_label_C42B_ER <- bind_rows(
  top_genes_C42B_ER_shMECOM,  # Top 5 up and downregulated genes
  genes_C42B_ER  # Only include genes of interest if they exist in the dataset
) %>%
  dplyr::distinct(symbol, .keep_all = TRUE)  

ggplot(data %>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")), aes(logFC, -log(PValue,10))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("logFC")) + 
  ylab(expression("-log"[10]*"pvalue")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick4")) +
  guides(colour = guide_legend(override.aes = list(size=1.5)))+
  ggrepel::geom_text_repel(data = genes_to_label_C42B_ER%>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")), mapping = aes(label = symbol),
                   size = 2.5, nudge_x=1, segment.size=0.25, direction="y",
                   fontface = "bold", max.overlaps = Inf)+
  geom_vline(xintercept=c(-0.6, 0.6), col="red", linetype="dashed") +
  geom_hline(yintercept=-log10(0.05), col="red", linetype="dashed")+
  ggtitle("C42B_ER_shMECOM_v_control")

# Count total, upregulated, and downregulated genes
total_genes <- nrow(data %>% dplyr::filter(contrast_id == "C42B_ER_shMECOM_v_control"))
upregulated_genes <- nrow(data %>% dplyr::filter(contrast_id == "C42B_ER_shMECOM_v_control", logFC > 0.5))
downregulated_genes <- nrow(data %>% dplyr::filter(contrast_id == "C42B_ER_shMECOM_v_control", logFC < -0.5))

# Calculate percentages
upregulated_percentage <- (upregulated_genes / total_genes) * 100
downregulated_percentage <- (downregulated_genes / total_genes) * 100

# Print the percentages in the console
print(paste("Upregulated genes:", round(upregulated_percentage, 2), "%"))
print(paste("Downregulated genes:", round(downregulated_percentage, 2), "%"))

# Volcano plot with labels
C42B_ER_shMECOM <- ggplot(data %>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")), aes(logFC, -log10(PValue))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("logFC")) +
  ylab(expression("-log"[10]*"pvalue")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick4")) +
  guides(colour = guide_legend(override.aes = list(size=1.5))) +
  ggrepel::geom_text_repel(
    data = genes_to_label_C42B_ER %>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")),
    mapping = aes(label = symbol),
    size = 3.0, nudge_x = 1, segment.size = 0.25, direction = "y",
    fontface = "bold", max.overlaps = Inf
  ) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "red", linetype = "dashed") +   
geom_hline(yintercept = -log10(0.05), col = "red", linetype = "dashed") +      
annotate("text", x = 4.5, y = min(-log10(data$PValue)) + 6, 
         label = sprintf("Up-regulated:     %6d\nDown-regulated:%6d\nTotal genes:         %6d\np-value:               %6.2f\nq-value:               %6.2f\nlogFC:                  %6.2f", 
                         round(upregulated_genes, 2), 
                         round(downregulated_genes, 2), 
                         total_genes, 
                         round(p_value,2), 
                         round(q_value,2), 
                         round(logFC,2)), 
         color = "black", size = 3.0, hjust = 0, fontface = "bold") +
ggtitle(expression(C42B^"-ER"~"_shMECOM_v_control"))


# Save the plot as a PDF
ggsave("C42B_ER_shMECOM.pdf", plot = C42B_ER_shMECOM, width = 8, height = 10, device = "pdf")

# Save the filtered data as a CSV file
C42B_ER_shMECOM <- C42B_ER_shMECOM$data
C42B_ER_shMECOM %>% write_csv("C42B_ER_shMECOM.csv")
```


```{r}

p_value <- 0.05  
q_value <- 0.05  
logFC   <- 1.5

# Ensure genes of interest are present in `data`
genes_CWRR1 <- data %>%
  dplyr::filter(contrast_id == "CWRR1_shMECOM_v_control", symbol %in% genes_of_interest_CWRR1)

# Combine top genes and genes of interest
genes_to_label_CWRR1 <- bind_rows(
  top_genes_CWRR1_shMECOM,  
  genes_CWRR1  
) %>%
  dplyr::distinct(symbol, .keep_all = TRUE) 

ggplot(data %>% dplyr::filter(contrast_id %in% c("CWRR1_shMECOM_v_control")), aes(logFC, -log(PValue,10))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("logFC")) + 
  ylab(expression("-log"[10]*"pvalue")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick4")) +
  guides(colour = guide_legend(override.aes = list(size=1.5)))+
  ggrepel::geom_text_repel(data = genes_to_label_CWRR1, mapping = aes(label = symbol),
                           size = 2, nudge_x=1, segment.size=0.25, direction="both",
                   fontface = "bold")+
  geom_vline(xintercept=c(-0.6, 0.6), col="red", linetype="dashed") +
  geom_hline(yintercept=-log10(0.05), col="red", linetype="dashed")+
  ggtitle("CWR-R1_shMECOM_v_control")

# Count total, upregulated, and downregulated genes
total_genes_CWRR1 <- nrow(data %>% dplyr::filter(contrast_id == "CWRR1_shMECOM_v_control"))
upregulated_genes <- nrow(data %>% dplyr::filter(contrast_id == "CWRR1_shMECOM_v_control", logFC > 0.6))
downregulated_genes <- nrow(data %>% dplyr::filter(contrast_id == "CWRR1_shMECOM_v_control", logFC < -0.6))

# Calculate percentages
upregulated_percentage <- (upregulated_genes / total_genes_CWRR1) * 100
downregulated_percentage <- (downregulated_genes / total_genes_CWRR1) * 100

# Print the percentages in the console
print(paste("Upregulated genes:", round(upregulated_percentage, 2), "%"))
print(paste("Downregulated genes:", round(downregulated_percentage, 2), "%"))

# Volcano plot with labels
CWRR1_shMECOM <- ggplot(data %>% dplyr::filter(contrast_id %in% c("CWRR1_shMECOM_v_control")), aes(logFC, -log10(PValue))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("logFC")) +
  ylab(expression("-log"[10]*"pvalue")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick4")) +
  guides(colour = guide_legend(override.aes = list(size=1.5))) +
  ggrepel::geom_text_repel(
    data = genes_to_label_CWRR1 %>% dplyr::filter(contrast_id %in% c("CWRR1_shMECOM_v_control")),
    mapping = aes(label = symbol),
    size = 2.5, nudge_x = 1, segment.size = 0.25, direction = "y",
    fontface = "bold", max.overlaps = Inf
  ) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "red", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "red", linetype = "dashed")+
  
  annotate("text", x = 4.5, y = min(-log10(data$PValue)) + 6, 
            label = sprintf("Up-regulated:     %6d\nDown-regulated:%6d\nTotal genes:         %6d\np-value:               %6.2f\nq-value:               %6.2f\nlogFC:                  %6.2f", 
                         round(upregulated_genes, 2), 
                         round(downregulated_genes, 2), 
                         total_genes, 
                         round(p_value,2), 
                         round(q_value,2), 
                         round(logFC,2)), 
         color = "black", size = 3.0, hjust = 0, fontface = "bold") +
 ggtitle(expression("CWR-R1_shMECOM_v_control"))

# Save the plot
ggsave("CWRR1_shMECOM.pdf", plot = CWRR1_shMECOM, width = 8, height = 10, device = "pdf")
```


```{r}
p_value <- 0.05  
q_value <- 0.05  
logFC   <- 1.5

#  Combine with specific genes of interest
genes_of_interest_LREX <- c("PBX1", "MAPK11", "FOXA1", "NOTCH1", "GATA2", "AR", "MBNL2")

# Ensure genes of interest are present in `data`
genes_LREX <- data %>%
  dplyr::filter(contrast_id == "LREX_shMECOM_v_control", symbol %in% genes_of_interest_LREX)

# Combine top genes and genes of interest
genes_to_label_LREX <- bind_rows(
  top_genes_LREX_shMECOM,  
  genes_LREX  
) %>%
  dplyr::distinct(symbol, .keep_all = TRUE) 

ggplot(data %>% dplyr::filter(contrast_id %in% c("LREX_shMECOM_v_control")), aes(logFC, -log(PValue,10))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("logFC")) + 
  ylab(expression("-log"[10]*"pvalue")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick4")) +
  guides(colour = guide_legend(override.aes = list(size=1.5)))+
  ggrepel::geom_text_repel(data = genes_to_label_LREX, mapping = aes(label = symbol),
                           size = 2.5, nudge_x=1, segment.size=0.25, direction="both",
                           fontface = "bold")+
  geom_vline(xintercept=c(-0.6, 0.6), col="red", linetype="dashed") +
  geom_hline(yintercept=-log10(0.05), col="red", linetype="dashed")+
  ggtitle("LREX_shMECOM_v_control")

#  Combine with specific genes of interest
genes_of_interest_LREX <- c("PBX1", "MAPK11", "FOXA1", "NOTCH1", "GATA2", "AR", "MBNL2")

# Count total, upregulated, and downregulated genes
total_genes_LREX <- nrow(data %>% dplyr::filter(contrast_id == "LREX_shMECOM_v_control"))
upregulated_genes <- nrow(data %>% dplyr::filter(contrast_id == "LREX_shMECOM_v_control", logFC > 0.6))
downregulated_genes <- nrow(data %>% dplyr::filter(contrast_id == "LREX_shMECOM_v_control", logFC < -0.6))

# Calculate percentages
upregulated_percentage <- (upregulated_genes / total_genes_LREX) * 100
downregulated_percentage <- (downregulated_genes / total_genes_LREX) * 100

# Print the percentages in the console
print(paste("Upregulated genes:", round(upregulated_percentage, 2), "%"))
print(paste("Downregulated genes:", round(downregulated_percentage, 2), "%"))

# Volcano plot with labels
LREX_shMECOM <- ggplot(data %>% dplyr::filter(contrast_id %in% c("LREX_shMECOM_v_control")), aes(logFC, -log10(PValue))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("logFC")) +
  ylab(expression("-log"[10]*"pvalue")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick4")) +
  guides(colour = guide_legend(override.aes = list(size=1.5))) +
  ggrepel::geom_text_repel(
    data = genes_to_label_LREX %>% dplyr::filter(contrast_id %in% c("LREX_shMECOM_v_control")),
    mapping = aes(label = symbol),
    size = 2.5, nudge_x = 1, segment.size = 0.25, direction = "y",
    fontface = "bold", max.overlaps = Inf
  ) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "red", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "red", linetype = "dashed") +
  annotate("text", x = 4.5, y = min(-log10(data$PValue)) + 8, 
            label = sprintf("Up-regulated:     %6d\nDown-regulated:%6d\nTotal genes:         %6d\np-value:               %6.2f\nq-value:               %6.2f\nlogFC:                  %6.2f", 
                         round(upregulated_genes, 2), 
                         round(downregulated_genes, 2), 
                         total_genes, 
                         round(p_value,2), 
                         round(q_value,2), 
                         round(logFC,2)), 
         color = "black", size = 3.0, hjust = 0, fontface = "bold") +
  ggtitle("LREX_shMECOM_v_control")

# Save the plot 
ggsave("LREX_shMECOM.pdf", plot = LREX_shMECOM, width = 8, height = 10, device = "pdf")
```


```{r}
library(ggVennDiagram)
library(VennDiagram)

C42B_ER_up <- data%>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")) %>% 
  dplyr::filter(Expression == 'Up-regulated')
C42B_ER_down <- data %>% dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control")) %>%
  dplyr::filter(Expression == 'Down-regulated')

CWRR1_up <- data%>% dplyr::filter(contrast_id %in% c("CWRR1_shMECOM_v_control")) %>% 
  dplyr::filter(Expression == 'Up-regulated')
CWRR1_down <- data %>% dplyr::filter(contrast_id %in% c("CWRR1_shMECOM_v_control")) %>%
  dplyr::filter(Expression == 'Down-regulated')

LREX_up <- data%>% dplyr::filter(contrast_id %in% c("LREX_shMECOM_v_control")) %>% 
  dplyr::filter(Expression == 'Up-regulated')
LREX_down <- data %>% dplyr::filter(contrast_id %in% c("LREX_shMECOM_v_control")) %>%
  dplyr::filter(Expression == 'Down-regulated')

venn_up <- list(C42B_ER_shMECOM_v_control = c(C42B_ER_up$symbol),
                LREX_shMECOM_v_control = c(LREX_up$symbol),
                CWRR1_shMECOM_v_control = c(CWRR1_up$symbol))

venn_down <- list(C42B_ER_shMECOM_v_control = c(C42B_ER_down$symbol),
               LREX_shMECOM_v_control = c(LREX_down$symbol),
                CWRR1_shMECOM_v_control = c(CWRR1_down$symbol)) 

ggVennDiagram(venn_up, set_color = c("darkgoldenrod4","firebrick4", "orangered"),
              label = "count", label_alpha = 0, set_size = 4, face="bold")+
  scale_x_continuous(expand = expansion(mult = .5))+
  scale_fill_distiller(palette = "Spectral")+
  ggtitle("Overlap of Significantly Up-regulated Genes")+
  theme(plot.title = element_text(face="bold", hjust = 0.5))
ggsave("Up_regulated.pdf", width = 8, height = 8)
ggVennDiagram(venn_down, set_color = c("darkgoldenrod4","firebrick4", "orangered"),
              label = "count", label_alpha = 0, set_size = 4, face="bold")+
  scale_x_continuous(expand = expansion(mult = .5))+
  scale_fill_distiller(palette = "Spectral")+
  ggtitle("Overlap of Significantly down-regulated Genes")+
  theme(plot.title = element_text(face="bold", hjust = 0.5))
ggsave("Down_regulated.pdf", width = 8, height = 8)
# overlapped genes
d1 <- calculate.overlap(venn_up)
d1[["a31"]]
d1[["a30"]]
d1[["a29"]]
d2 <- calculate.overlap(venn_down)
d2[["a31"]]
```
```{r}
library(msigdbr)          
library(clusterProfiler) 
library(purrr)           
library(data.table)       
library(AnnotationDbi)  
library(org.Hs.eg.db)    

# Importing all gene-sets
msigdb <- msigdbr() 
AR_Targets <- list("non-canonical AR" = data.frame(gene = c("CDC20P1", "UBE2C", "CDK1", "BCAT1", "SPG20", "CNTNAP4", "GJB2", "ELF5", "CCNA2", "MEI1", "TERF1", "CD74", "ANAPC10", "CDH3", "CDK5", "TOP2B", "PRDM4", "CYP7A1", "FGF21", "KRT2", "MET", "CD74", "MBNL2", "RNF165", "CREB5", "TGFBI", "SEC61A1", "AGR2", "ABCG2", "GATA2", "FOLH1")),
                   "canonical AR" = data.frame(gene = c("KLK3", "KLK2", "FKBP5", "STEAP1", "STEAP2", "STEAP4", "STEAP3", "PPAP2A", "RAB3B", "ACSL3", "NKX3-1", "KLK4", "KLK15", "TMPRSS2", "DDR2", "ZBTB16", "CDK6", "KRT19", "ORM1","POU2F1")))
AR_genesets <- AR_Targets %>% 
  bind_rows(.id = "ont")
all_sets <- msigdb %>% 
  transmute(ont = gs_name,
            gene = gene_symbol) %>% bind_rows(AR_genesets)
sorted_data <- topTable[,c("symbol", "logFC", "contrast_id")] %>% arrange(desc(logFC)) 

selected_contrast_ids <- c("C42B_ER_shMECOM_v_control", "CWRR1_shMECOM_v_control", "LREX_shMECOM_v_control")

filtered_sorted_data <- sorted_data %>%
  dplyr::filter(contrast_id %in% selected_contrast_ids)

gsea_results <- lapply(unique(filtered_sorted_data$contrast_id), function(contrast){
  contrast_stats <- filtered_sorted_data$logFC[filtered_sorted_data$contrast_id == contrast]
  names(contrast_stats) <- filtered_sorted_data$symbol[filtered_sorted_data$contrast_id == contrast]
  gsea <- GSEA(contrast_stats, 
               TERM2GENE = all_sets, 
               pvalueCutoff = 1)
  return(gsea)
}) %>%
  setNames(unique(filtered_sorted_data$contrast_id))
gsea_Annotated <- gsea_results %>% lapply(function(x){return(x@result)}) %>% bind_rows(.id = "contrast_id")
gsea_Annotated <- gsea_Annotated %>% left_join(msigdb %>% distinct(gs_name,gs_cat, gs_subcat), by = c("ID"="gs_name"))
gsea_Annotated <- gsea_Annotated %>% mutate(
  Enrichment = case_when(NES >= 0.1 & qvalue <= 0.05 ~ "Enrichment-Up",
                         NES <= -0.1 & qvalue <= 0.05 ~ "Enrichment-Down",
                         TRUE ~ "Unchanged")
)
```


```{r}
gsea_Annotated <- as.data.frame(gsea_Annotated)
GSEA_data <- gsea_Annotated 

ggplot(gsea_Annotated %>% dplyr::filter(ID %in% c("WANG_PROSTATE_CANCER_ANDROGEN_INDEPENDENT", "non-canonical AR",
                                                  "PID_AR_PATHWAY", "PID_AR_NONGENOMIC_PATHWAY", "NELSON_RESPONSE_TO_ANDROGEN_UP", "NELSON_RESPONSE_TO_ANDROGEN_DN",
                                                  "HALLMARK_ANDROGEN_RESPONSE", "GO_NEGATIVE_REGULATION_OF_ANDROGEN_RECEPTOR_SIGNALING_PATHWAY", "canonical AR") & contrast_id %in% c("LREX_shMECOM_v_control", "C42B_ER_shMECOM_v_control", "CWRR1_shMECOM_v_control")), 
       aes(x = contrast_id, y = ID, size = -log10(pvalue), color = NES)) +
  geom_point() +
  scale_colour_gradient2( low= "blue4",  high= "red4") +
  ylab("")+
  xlab("")+
  theme_gray(base_size = 12) +
  ggtitle("GSEA: AR PATHWAYS") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
  ggsave("GSEA_AR_PATHWAYS.pdf", width = 8, height = 8)
```


```{r}
gsea_plot_data <- gsea_Annotated %>%
  dplyr::filter(contrast_id %in% c("C42B_ER_shMECOM_v_control", "CWRR1_shMECOM_v_control","LREX_shMECOM_v_control"))

ggplot(gsea_plot_data, 
       aes(x = contrast_id, y = ID, size = -log10(pvalue), color = NES)) +
  geom_point() +
  scale_colour_gradient2(low = "blue4", mid = "white", high = "red4", midpoint = 0) +
  scale_size_continuous(range = c(3, 10)) + 
  ylab(" ") +
  xlab(" ") +
  theme_gray(base_size = 12) +
  ggtitle("GSEA: ") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, size = 10, margin = margin(t = 10)),
    axis.text.y = element_text(size = 8, margin = margin(r = 10)),
    plot.title = element_text(hjust = 0.5)
  ) 

ggsave("ALL_PATHWAYS.pdf", width = 8, height = 24)  # Increased height for better clarity

write_csv(gsea_plot_data, "Enriched_Genesets_All_ContrastIDs.csv")
```


```{r}
shMECOM_enriched_genesets <- gsea_Annotated %>%
  dplyr::filter(grepl("shMECOM", contrast_id) &   # Filter for shMECOM contrasts
         (abs(NES) > 0 & pvalue < 0.05) &
           contrast_id %in% selected_contrast_ids &
           
           Enrichment %in% c("Enrichment-Up", "Enrichment-Down"))  %>%
  distinct(contrast_id, ID, .keep_all = TRUE)


write_csv(shMECOM_enriched_genesets, "shMECOM_enriched_genesets.csv" )

selected_contrast_ids <- c("CWRR1_shMECOM_v_control", "lREX_shMECOM_v_control", "C42B_ER_shMECOM_v_control")

shMECOM_enriched_genesets_UP <- gsea_Annotated %>%
  dplyr::filter(grepl("shMECOM", contrast_id) &            
         NES > 0 &                                 
         pvalue < 0.05 &                            
         contrast_id %in% selected_contrast_ids &
  Enrichment == "Enrichment-Up" ) %>% 
  distinct(contrast_id, ID, .keep_all = TRUE)   


shMECOM_genesets_UP <- split(shMECOM_enriched_genesets_UP, shMECOM_enriched_genesets_UP$contrast_id)
shMECOM_geneset_list <- lapply(shMECOM_genesets_UP, function(df) df$ID)

enriched_overlapping_genesets <- Reduce(intersect, shMECOM_geneset_list)

Enriched_UP_overlapping_genesets_data <- shMECOM_enriched_genesets_UP %>%
  dplyr::filter(ID %in% enriched_overlapping_genesets)

print(Enriched_UP_overlapping_genesets_data)

write_csv(Enriched_UP_overlapping_genesets_data, "Enriched_UP_overlapping_enriched_genesets.csv")
```


```{r}
shMECOM_enriched_genesets_Down <- gsea_Annotated %>%
  dplyr::filter(grepl("shMECOM", contrast_id) &            
         NES < 0 &                                 
         pvalue < 0.05 &                           
         contrast_id %in% selected_contrast_ids &
  
  Enrichment == "Enrichment-Down") %>% 
  
  distinct(contrast_id, ID, .keep_all = TRUE)  

shMECOM_genesets_Down <- split(shMECOM_enriched_genesets_Down, shMECOM_enriched_genesets_Down$contrast_id)
shMECOM_geneset_list_Down <- lapply(shMECOM_genesets_Down, function(df) df$ID)

enriched_Down_overlapping_genesets <- Reduce(intersect, shMECOM_geneset_list_Down)

Enriched_Down_overlapping_genesets_data <- shMECOM_enriched_genesets_Down %>%
  dplyr::filter(ID %in% enriched_Down_overlapping_genesets)

print(Enriched_Down_overlapping_genesets_data)

write_csv(Enriched_Down_overlapping_genesets_data, "Enriched_Down_overlapping_enriched_genesets.csv")
```


```{r}
# Define your custom gene set
AR_Targets <- list(
  "non-canonical AR" = data.frame(gene = c("CDC20P1", "UBE2C", "CDK1", "BCAT1", "SPG20", "CNTNAP4", "GJB2", "ELF5", "CCNA2", "MEI1", "TERF1", "CD74", "ANAPC10", "CDH3", "CDK5", "TOP2B", "PRDM4", "CYP7A1", "FGF21", "KRT2", "MET", "CD74", "MBNL2", "RNF165", "CREB5", "TGFBI", "SEC61A1", "AGR2", "ABCG2", "GATA2", "FOLH1")),
  "canonical AR" = data.frame(gene = c("KLK3", "KLK2", "FKBP5", "STEAP1", "STEAP2", "STEAP4", "STEAP3", "PPAP2A", "RAB3B", "ACSL3", "NKX3-1", "KLK4", "KLK15", "TMPRSS2", "DDR2", "ZBTB16", "CDK6", "KRT19", "ORM1", "POU2F1"))
)

if (!exists("gsea_results")) {
  stop("Error: gsea_results object not found! Please check your GSEA results file.")
}

enrich_plot_data <- lapply(gsea_results, enrichplot:::gsInfo, "non-canonical AR") %>% 
  bind_rows(.id = "contrast_id") %>%
  dplyr::filter(!is.na(x)) 

num_contrasts <- length(unique(enrich_plot_data$contrast_id))
phenotypecolors <- colorRampPalette(c("red", "darkblue", "orange"))(num_contrasts)

p1 <- ggplot(enrich_plot_data, aes(x = x, y = runningScore, color = contrast_id)) + 
  geom_line(size = 1) +  # Make the line thicker for visibility
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.7) +  # Add zero-line
  scale_color_manual(values = phenotypecolors) + 
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        legend.position = "none") +  # Remove legend
  ylab("Enrichment Score") + 
  xlab("Position in the Ranked List of Genes") + 
  ggtitle("Hallmark: Androgen Response")

gene_hits <- enrich_plot_data %>%
  dplyr::filter(gene %in% AR_Targets[["non-canonical AR"]]$gene) %>%  # Use your custom gene set
  dplyr::select(x, contrast_id) %>% 
  drop_na(x)


if (nrow(gene_hits) == 0) {
  warning("No gene hits found for the selected geneset. Check GSEA results.")
} else {
  print("Gene hits found. Proceeding to plot.")
}

# **Gene Hit Visualization - Rug Plot**
p2 <- ggplot(gene_hits, aes(x = x, color = contrast_id)) + 
  geom_rug(sides = "b", size = 0.7, alpha = 0.7) +  
  scale_color_manual(values = phenotypecolors) +
  theme_minimal() +  
  theme(axis.title.x = element_blank(), 
        axis.title.y = element_blank(),
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        legend.position = "none",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        plot.background = element_blank())

final_plot <- grid.arrange(p1, p2, ncol = 1, heights = c(3, 1)) 
```





